qiskit/quantum_info/operators/dihedral/dihedral.py

```diff
@@ -0,0 +1,495 @@
+   (x_0,x_0 + x_1,x_2 + 1)
+
+ This means that |x_0 x_1 x_2> transforms to omega^{p(x)}|f(x)>,
+ where omega = exp(i*pi/4) from which we can read that
+ T^3 on qubit 1 AFTER CNOT_{0,1} is the same as
+ T^3 on qubit 0, T^3 on qubit 1, and CS_{0,1} BEFORE CNOT_{0,1}.
+"""
+
+import itertools
+import numpy as np
+
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit.quantum_info.operators.operator import Operator
+from qiskit.quantum_info.operators.pauli import Pauli
+from qiskit.quantum_info.operators.scalar_op import ScalarOp
+from qiskit.quantum_info.synthesis.cnotdihedral_decompose import decompose_cnotdihedral
+from qiskit.circuit import QuantumCircuit, Instruction
+from .dihedral_circuits import _append_circuit
+from .polynomial import SpecialPolynomial
+
+
+class CNOTDihedral(BaseOperator):
+    """CNOT-dihedral Object Class.
+    The CNOT-dihedral group on num_qubits qubits is generated by the gates
+    CNOT, T and X.
+
+    References:
+        1. Shelly Garion and Andrew W. Cross, *Synthesis of CNOT-Dihedral circuits
+           with optimal number of two qubit gates*, `Quantum 4(369), 2020
+           <https://quantum-journal.org/papers/q-2020-12-07-369/>`_
+        2. Andrew W. Cross, Easwar Magesan, Lev S. Bishop, John A. Smolin and Jay M. Gambetta,
+           *Scalable randomised benchmarking of non-Clifford gates*,
+           npj Quantum Inf 2, 16012 (2016).
+    """
+
+    def __init__(self, data, validate=True):
+        """Initialize a CNOTDihedral operator object."""
+
+        # Initialize from another CNOTDihedral by sharing the underlying
+        # poly, linear and shift
+        if isinstance(data, CNOTDihedral):
+            self.linear = data.linear
+            self.shift = data.shift
+            self.poly = data.poly
+
+        # Initialize from ScalarOp as N-qubit identity discarding any global phase
+        elif isinstance(data, ScalarOp):
+            if not data.is_unitary() or set(data._input_dims) != {2} or \
+                    data.num_qubits is None:
+                raise QiskitError("Can only initialize from N-qubit identity ScalarOp.")
+            self._num_qubits = data.num_qubits
+            # phase polynomial
+            self.poly = SpecialPolynomial(self._num_qubits)
+            # n x n invertible matrix over Z_2
+            self.linear = np.eye(self._num_qubits, dtype=np.int8)
+            # binary shift, n coefficients in Z_2
+            self.shift = np.zeros(self._num_qubits, dtype=np.int8)
+
+        # Initialize from a QuantumCircuit or Instruction object
+        elif isinstance(data, (QuantumCircuit, Instruction)):
+            self._num_qubits = data.num_qubits
+            elem = self.from_circuit(data)
+            self.poly = elem.poly
+            self.linear = elem.linear
+            self.shift = elem.shift
+
+        # Construct the identity element on num_qubits qubits.
+        elif isinstance(data, int):
```

Comment:
I would remove this identity init for now.

If you want identity you could do init as:
```python
def __init__(data=None, num_qubits=None, validate=True):
    if num_qubits:
        # initialize n-qubit identity
    elif isinstance(data, ...):
        # other cases
```

This is clearer I think and you can document it in the init docs under args.